zipline/finance/controls.py

```diff
@@ -286,12 +288,34 @@ def validate(self,
         Fail if the algo has passed this Asset's end_date, or before the
         Asset's start date.
         """
+        normalized_algo_dt = pd.Timestamp(algo_datetime).normalize()
+
+        # If the order is for 0 shares, then silently pass through.
+        if amount == 0:
+            return
+
         # Fail if the algo is before this Asset's start_date
-        if asset.start_date and (algo_datetime < asset.start_date):
-            self.fail(asset, amount, algo_datetime)
+        if asset.start_date:
+            normalized_start = pd.Timestamp(asset.start_date).normalize()
+            if normalized_algo_dt < normalized_start:
+                self._fail_name = "asset_start_date"
+                self._fail_date = normalized_start
+                self.fail(asset, amount, algo_datetime)
         # Fail if the algo has passed this Asset's end_date
-        if asset.end_date and (algo_datetime >= asset.end_date):
-            self.fail(asset, amount, algo_datetime)
+        if asset.end_date:
+            normalized_end = pd.Timestamp(asset.end_date).normalize()
+            if normalized_algo_dt > normalized_end:
+                self._fail_name = "asset_end_date"
+                self._fail_date = normalized_end
+                self.fail(asset, amount, algo_datetime)
+
+    def __repr__(self):
+        # More descriptive repr, since AssetDateBounds depend on a dynamic
+        # comparison
+        return "{name}(\"{fail_name}\":\"{fail_date})\"".format(
+            name=self.__class__.__name__,
+            fail_name=self._fail_name,
```

Comment:
It's great that this is more descriptive, but it's also confusing to me that I can't instantiate an AssetDateBounds and get its repr without having called validate first.  Also, if I call validate again, I'll see a different repr.  I think that a repr usually represents how you might instantiate this object, which this also fails on (no pun intended).

When are we looking at this repr?  Should whatever ```fail``` returns or raises have a better repr instead?